python/cudf/cudf/core/indexed_frame.py

```diff
@@ -1288,6 +1289,93 @@ def _align_to_index(
 
         return result
 
+    @_cudf_nvtx_annotate
+    def _reindex(
+        self,
+        columns,
+        dtypes=None,
+        deep=False,
+        index=None,
+        inplace=False,
+        fill_value=NA,
+    ):
+        """
+        Helper for `.reindex`
+
+        Parameters
+        ----------
+        columns : array-like
+            The list of columns to select from the Frame,
+            if ``columns`` is a superset of ``Frame.columns`` new
+            columns are created.
+        dtypes : dict
+            Mapping of dtypes for the empty columns being created.
+        deep : boolean, optional, default False
+            Whether to make deep copy or shallow copy of the columns.
+        index : Index or array-like, default None
+            The ``index`` to be used to reindex the Frame with.
+        inplace : bool, default False
+            Whether to perform the operation in place on the data.
+        fill_value : value with which to replace nulls in the result
+
+        Returns
+        -------
+        DataFrame
+        """
+        if dtypes is None:
+            dtypes = {}
+
+        df = self
+        if index is not None:
+            index = cudf.core.index.as_index(index)
+
+            idx_dtype_match = (df.index.nlevels == index.nlevels) and all(
+                left_dtype == right_dtype
+                for left_dtype, right_dtype in zip(
+                    (col.dtype for col in df.index._data.columns),
+                    (col.dtype for col in index._data.columns),
+                )
+            )
+
+            if not idx_dtype_match:
+                columns = (
+                    columns if columns is not None else list(df._column_names)
+                )
+                df = cudf.DataFrame()
+            else:
+                df = cudf.DataFrame(None, index).join(
+                    cudf.DataFrame(df), how="left", sort=True
+                )
```

Comment:
Can we avoid using public constructors and instead use something like `cudf.DataFrame._from_data()` instead? The constructors

Reply:
I think the df (an empty dataframe with index) can be constructed by
```
cudf.DataFrame._from_data({}, index=index)
```
The second one is an `indexed_frame` right? so perhaps just use df directly?

Reply:
I updated this to use `_from_data`. The RHS unfortunately still needs to be cast to a `DataFrame` for all the bookkeeping in `_merge` to work out it seems, but I was able to still remove the public constructor.